aggregator/flush_mgr.go

```diff
@@ -0,0 +1,338 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errFlushManagerClosed = errors.New("flush manager is closed")
+)
+
+// PeriodicFlusher flushes periodically.
+type PeriodicFlusher interface {
+	// FlushInterval returns the periodic flush interval.
+	FlushInterval() time.Duration
+
+	// Flush performs a flush.
+	Flush()
+}
+
+// FlushManager manages and coordinates flushing activities across many
+// periodic flushers with different flush intervals with controlled concurrency
+// for flushes to minimize spikes in CPU load and reduce p99 flush latencies.
+type FlushManager interface {
+	// Register registers a metric list with the flush manager.
+	Register(flusher PeriodicFlusher) error
+
+	// Close closes the flush manager.
+	Close()
+}
+
+type flushManagerMetrics struct {
+	queueSize tally.Gauge
+}
+
+func newFlushManagerMetrics(scope tally.Scope) flushManagerMetrics {
+	return flushManagerMetrics{
+		queueSize: scope.Gauge("queue-size"),
+	}
+}
+
+type flushManager struct {
+	sync.Mutex
+
+	nowFn         clock.NowFn
+	checkEvery    time.Duration
+	jitterEnabled bool
+	workers       xsync.WorkerPool
+	scope         tally.Scope
+
+	closed     bool
+	rand       *rand.Rand
+	buckets    []*flushBucket
+	flushTimes flushMetadataHeap
+	sleepFn    sleepFn
+	wgFlush    sync.WaitGroup
+	metrics    flushManagerMetrics
+}
+
+// NewFlushManager creates a new flush manager.
+func NewFlushManager(opts FlushManagerOptions) FlushManager {
+	if opts == nil {
+		opts = NewFlushManagerOptions()
+	}
+	nowFn := opts.ClockOptions().NowFn()
+	scope := opts.InstrumentOptions().MetricsScope()
+	rand := rand.New(rand.NewSource(nowFn().UnixNano()))
+	workers := xsync.NewWorkerPool(opts.WorkerPoolSize())
+	workers.Init()
+
+	mgr := &flushManager{
+		nowFn:         nowFn,
+		checkEvery:    opts.CheckEvery(),
+		jitterEnabled: opts.JitterEnabled(),
+		workers:       workers,
+		scope:         scope,
+		rand:          rand,
+		sleepFn:       time.Sleep,
+		metrics:       newFlushManagerMetrics(scope),
+	}
+	if mgr.checkEvery > 0 {
+		mgr.wgFlush.Add(1)
+		go mgr.flush()
+	}
+	return mgr
+}
+
+func (mgr *flushManager) Register(flusher PeriodicFlusher) error {
+	mgr.Lock()
+	bucket, err := mgr.bucketForWithLock(flusher)
+	if err == nil {
+		bucket.Add(flusher)
```

Comment:
this means if the flusher has a one min interval and it registered it self 10 seconds after the last one min bucket was flushed, this flusher maybe flushed in 50 second later? is that ok?